File: src/px4_ros_com/src/image_stitching_node.cpp
// ----------------------------------------------------
// Image Stitcher Node (ROS2 + OpenCV)
// ----------------------------------------------------
// - Subscribes to a ROS2 image topic
// - Buffers incoming images
// - Every N seconds, attempts to stitch them together
// - Publishes and saves the stitched result
// ----------------------------------------------------

#include <rclcpp/rclcpp.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/stitching.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/msg/image.hpp>

#include <vector>
#include <string>
#include <filesystem>
#include <chrono>
#include <iomanip>

// ------------------------------
// Image Stitcher Node Definition
// ------------------------------
class ImageStitcherNode : public rclcpp::Node {
public:
  ImageStitcherNode()
      : Node("image_stitcher_node") {
    
    // Declare and load ROS parameters.
    this->declare_parameter<std::string>("output_path", "/tmp/stitched.jpg");
    this->declare_parameter<bool>("crop", true);
    this->declare_parameter<bool>("preprocessing", false);
    this->declare_parameter<double>("stitch_interval_sec", 10.0);  // How often to stitch (in seconds)

    this->get_parameter("output_path", output_path_);
    this->get_parameter("crop", crop_);
    this->get_parameter("preprocessing", preprocessing_);
    this->get_parameter("stitch_interval_sec", stitch_interval_sec_);

    // Publisher to output stitched image
    image_pub_ = this->create_publisher<sensor_msgs::msg::Image>("stitched_image", 10);

    // Subscriber to image feed
    image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
      "/world/default/model/x500_mono_cam_down_0/link/camera_link/sensor/imager/image", 
      10,
      std::bind(&ImageStitcherNode::image_callback, this, std::placeholders::_1)
    );

    // Timer to stitch every N seconds
    stitch_timer_ = this->create_wall_timer(
      std::chrono::duration<double>(stitch_interval_sec_),
      std::bind(&ImageStitcherNode::timer_callback, this)
    );
  }

private:
  // Parameters
  std::string output_path_;
  bool crop_;
  bool preprocessing_;
  double stitch_interval_sec_;

  // ROS Interfaces
  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
  rclcpp::TimerBase::SharedPtr stitch_timer_;

  // Image buffer (stores images between stitching intervals)
  std::vector<cv::Mat> received_images_;

  // ------------------------------
  // Image Callback (Triggered on Every Frame)
  // ------------------------------
  void image_callback(const sensor_msgs::msg::Image::SharedPtr msg) {
    cv_bridge::CvImagePtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
    } catch (cv_bridge::Exception &e) {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
      return;
    }

    // Store the incoming image
    received_images_.push_back(cv_ptr->image.clone());
    RCLCPP_INFO(this->get_logger(), "Received image. Buffer size: %zu", received_images_.size());
  }

  // ------------------------------
  // Timer Callback (Triggered Every N Seconds)
  // ------------------------------
  void timer_callback() {
    if (received_images_.size() < 2) {
      RCLCPP_WARN(this->get_logger(), "Not enough images to stitch yet.");
      return;
    }

    RCLCPP_INFO(this->get_logger(), "Attempting to stitch %zu images...", received_images_.size());
    stitch_and_publish(received_images_);
    received_images_.clear();  // Clear buffer after stitching
  }

  // ------------------------------
  // Core Stitching and Publishing Logic
  // ------------------------------
  void stitch_and_publish(const std::vector<cv::Mat>& input_images) {
    auto resized = resize_images(input_images);
    auto preprocessed = preprocess_images(resized);
    cv::Mat stitched;
    auto status = stitch_images(preprocessed, stitched);

    if (status == cv::Stitcher::OK) {
      RCLCPP_INFO(this->get_logger(), "Image stitching successful.");
      crop_image(stitched);  // Optional post-processing crop

      // Generate timestamped filename
      auto now = std::chrono::system_clock::now();
      auto time = std::chrono::system_clock::to_time_t(now);
      std::stringstream ss;
      ss << std::put_time(std::localtime(&time), "stitched_%Y%m%d_%H%M%S.jpg");

      std::filesystem::path output_path_base(output_path_);
      std::string filename = (output_path_base.parent_path() / ss.str()).string();

      // Save to disk
      if (!cv::imwrite(filename, stitched)) {
        RCLCPP_ERROR(this->get_logger(), "Failed to save stitched image to %s", filename.c_str());
      } else {
        RCLCPP_INFO(this->get_logger(), "Saved stitched image to %s", filename.c_str());
      }

      // Publish over ROS
      auto msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", stitched).toImageMsg();
      image_pub_->publish(*msg);
      RCLCPP_INFO(this->get_logger(), "Published stitched image to topic: /stitched_image");
    } else {
      RCLCPP_ERROR(this->get_logger(), "Image stitching failed with status code: %d", static_cast<int>(status));
    }
  }

  // ------------------------------
  // Image Preprocessing Helpers
  // ------------------------------

  // Resize images to a consistent width if needed
  std::vector<cv::Mat> resize_images(const std::vector<cv::Mat>& images, int widthThreshold = 1500) {
    std::vector<cv::Mat> resized_images;
    for (const auto& image : images) {
      if (image.cols > widthThreshold) {
        double ratio = static_cast<double>(widthThreshold) / image.cols;
        cv::Size dim(widthThreshold, static_cast<int>(image.rows * ratio));
        cv::Mat resized;
        cv::resize(image, resized, dim);
        resized_images.push_back(resized);
      } else {
        resized_images.push_back(image);
      }
    }
    return resized_images;
  }

  // We can get rid of this, not sure if its better to have or leave out
  // Contrast enhancement (CLAHE)
  std::vector<cv::Mat> preprocess_images(std::vector<cv::Mat>& images) {
    if (!preprocessing_) return images;

    RCLCPP_INFO(this->get_logger(), "Preprocessing images...");
    for (auto& image : images) {
      cv::Mat lab_image;
      cv::cvtColor(image, lab_image, cv::COLOR_BGR2Lab);
      std::vector<cv::Mat> planes(3);
      cv::split(lab_image, planes);
      auto clahe = cv::createCLAHE(2.0, cv::Size(8, 8));
      clahe->apply(planes[0], planes[0]);
      cv::merge(planes, lab_image);
      cv::cvtColor(lab_image, image, cv::COLOR_Lab2BGR);
    }
    return images;
  }

  // Run OpenCV stitching pipeline
  cv::Stitcher::Status stitch_images(const std::vector<cv::Mat>& images, cv::Mat& stitched) {
    RCLCPP_INFO(this->get_logger(), "Stitching images...");
    cv::Ptr<cv::Stitcher> stitcher = cv::Stitcher::create(cv::Stitcher::SCANS);
    return stitcher->stitch(images, stitched);
  }

  // Auto-crop black borders from the stitched result
  void crop_image(cv::Mat& image) {
    if (!crop_) return;

    RCLCPP_INFO(this->get_logger(), "Cropping stitched image...");
    cv::copyMakeBorder(image, image, 10, 10, 10, 10, cv::BORDER_CONSTANT, cv::Scalar(0, 0, 0));

    cv::Mat gray, thresh;
    cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    cv::threshold(gray, thresh, 0, 255, cv::THRESH_BINARY | cv::THRESH_OTSU);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(thresh, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    if (!contours.empty()) {
      size_t maxIdx = 0;
      double maxArea = 0;
      for (size_t i = 0; i < contours.size(); ++i) {
        double area = cv::contourArea(contours[i]);
        if (area > maxArea) {
          maxArea = area;
          maxIdx = i;
        }
      }

      cv::Rect bounding_rect = cv::boundingRect(contours[maxIdx]);
      image = image(bounding_rect).clone();
    }
  }
};

// ------------------------------
// Main Function
// ------------------------------
int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImageStitcherNode>());
  rclcpp::shutdown();
  return 0;
}

File: src/px4_ros_com/src/image_stitching_node.py
#!/usr/bin/env python3

# ----------------------------------------------------
# Image Stitcher Node (ROS2 + OpenCV in Python)
# ----------------------------------------------------
# - Subscribes to a ROS2 image topic
# - Buffers incoming images
# - Every N seconds, stitches them
# - Saves and displays the result (no publishing)
# ----------------------------------------------------

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np
import os
from datetime import datetime

class ImageStitcherNode(Node):
    def __init__(self):
        super().__init__('image_stitcher_node')

        # Parameters
        self.declare_parameter('output_path', '/tmp/stitched.jpg')
        self.declare_parameter('crop', True)
        self.declare_parameter('preprocessing', False)
        self.declare_parameter('stitch_interval_sec', 3.0)

        self.output_path = self.get_parameter('output_path').get_parameter_value().string_value
        self.crop = self.get_parameter('crop').get_parameter_value().bool_value
        self.preprocessing = self.get_parameter('preprocessing').get_parameter_value().bool_value
        self.stitch_interval_sec = self.get_parameter('stitch_interval_sec').get_parameter_value().double_value

        self.image_sub = self.create_subscription(
            Image,
            '/world/baylands/model/x500_mono_cam_down_0/link/camera_link/sensor/imager/image',
            self.image_callback,
            10
        )

        self.stitch_timer = self.create_timer(self.stitch_interval_sec, self.timer_callback)

        self.bridge = CvBridge()
        self.received_images = []

        self.get_logger().info("Image Stitcher Node Initialized.")

    def image_callback(self, msg):
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msg, 'bgr8')
            self.received_images.append(cv_image)
            self.get_logger().info(f"Received image. Buffer size: {len(self.received_images)}")
        except Exception as e:
            self.get_logger().error(f"cv_bridge exception: {str(e)}")


    def timer_callback(self):
        if len(self.received_images) < 2:
            self.get_logger().warn("Not enough images to stitch yet.")

            return

        self.get_logger().info(f"Attempting to stitch {len(self.received_images)} images...")

        status, stitched = self.stitch_images(self.received_images)
        self.received_images.clear()

        if status == cv2.Stitcher_OK:
            self.get_logger().info("Image stitching successful.")

            if self.crop:
                stitched = self.crop_image(stitched)

            # Save stitched image
            now = datetime.now()
            timestamp = now.strftime("%Y%m%d_%H%M%S")
            output_file = os.path.join(os.path.dirname(self.output_path), f"stitched_{timestamp}.jpg")
            cv2.imwrite(output_file, stitched)
            self.get_logger().info(f"Stitched image saved to {output_file}")

            # Display stitched image
            cv2.imshow("Stitched Image", stitched)
            cv2.waitKey(1)  # 1ms non-blocking wait
        else:
            self.get_logger().error(f"Image stitching failed with status code: {status}")
            if status == cv2.Stitcher_ERR_NEED_MORE_IMGS:
                self.get_logger().error("Need more images to perform stitching.")
            elif status == cv2.Stitcher_ERR_HOMOGRAPHY_EST_FAIL:
                self.get_logger().error("Homography estimation failed. Improve overlap or reduce blur.")
            elif status == cv2.Stitcher_ERR_CAMERA_PARAMS_ADJUST_FAIL:
                self.get_logger().error("Camera parameter adjustment failed. Check capture settings.")

    def resize_images(self, images, widthThreshold=1500):
        resized_images = []
        for image in images:
            if image.shape[1] > widthThreshold:
                ratio = widthThreshold / image.shape[1]
                dim = (widthThreshold, int(image.shape[0] * ratio))
                resized_images.append(cv2.resize(image, dim))
            else:
                resized_images.append(image)
        return resized_images

    def preprocess_images(self, images):
        if not self.preprocessing:
            return images

        self.get_logger().info("[INFO] Preprocessing images to improve clarity...")
        for i in range(len(images)):
            lab = cv2.cvtColor(images[i], cv2.COLOR_BGR2LAB)
            l, a, b = cv2.split(lab)
            clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
            l = clahe.apply(l)
            lab = cv2.merge((l, a, b))
            images[i] = cv2.cvtColor(lab, cv2.COLOR_LAB2BGR)
        return images

    def stitch_images(self, images):
        images = self.resize_images(images)
        images = self.preprocess_images(images)
        stitcher = cv2.Stitcher.create(cv2.Stitcher_SCANS)
        return stitcher.stitch(images)

    def crop_image(self, stitched):
        self.get_logger().info("[INFO] Cropping stitched image...")
        stitched = cv2.copyMakeBorder(stitched, 10, 10, 10, 10,
                                      cv2.BORDER_CONSTANT, value=(0, 0, 0))
        gray = cv2.cvtColor(stitched, cv2.COLOR_BGR2GRAY)
        _, thresh = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)
        contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        if contours:
            c = max(contours, key=cv2.contourArea)
            x, y, w, h = cv2.boundingRect(c)
            stitched = stitched[y:y+h, x:x+w]
        return stitched


def main(args=None):
    rclpy.init(args=args)
    node = ImageStitcherNode()
    rclpy.spin(node)
    cv2.destroyAllWindows()
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
